text/0000-dst-coercion.md

```diff
@@ -0,0 +1,182 @@
+pub struct Rc<T: ?Sized> {
+    _ptr: NonZero<*mut RcBox<T>>,
+}
+pub struct NonZero<T: Zeroable+?Sized>(T);
+```
+
+Implementing `CoerceUnsized` indicates that the self type should be able to be
+coerced to the `Target` type. E.g., the above implementation means that
+`Rc<[i32; 42]>` can be coerced to `Rc<[i32]>`. There will be `CoerceUnsized` impls
+for the various pointer kinds available in Rust and which allow coercions, therefore
+`CoerceUnsized` when used as a bound indicates coercible types. E.g.,
+
+```
+fn foo<T: CoerceUnsized<U>, U>(x: T) -> U {
+    x
+}
+```
+
+Built-in pointer impls:
+
+```
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<Box<U>> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&'a U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&mut 'a U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*const U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*mut U> for Box<T> {}
+
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a, 'b: 'a> CoerceUnsized<&'a U> for &mut 'b U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&mut 'a U> for &mut 'a U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<*const U> for &mut 'a U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<*mut U> for &mut 'a U {}
+
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a, 'b> CoerceUnsized<&'a U> for &'b U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'b> CoerceUnsized<*const U> for &'b U {}
+
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*const U> for *mut U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*mut U> for *mut U {}
+
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*const U> for *const U {}
+```
+
+Note that there are some coercions which are not given by `CoerceUnsized`, e.g.,
+from safe to unsafe function pointers, so it really is a `CoerceUnsized` trait,
+not a general `Coerce` trait.
+
+
+## Compiler checking
+
+### On encountering an implementation of `CoerceUnsized` (type collection phase)
+
+* If the impl is for a built-in pointer type, we check nothing, otherwise...
+* The compiler checks that the `Self` type is a struct or tuple struct and that
+the `Target` type is a simple substitution of type parameters from the `Self`
+type (one day, with HKT, this could be a regular part of type checking, for now
+it must be an ad hoc check). We might enforce that this substitution is of the
+form `X/Y` where `X` and `Y` are both formal type parameters of the
+implementation (I don't think this is necessary, but it makes checking coercions
+easier and is satisfied for all smart pointers).
+* The compiler checks each field in the `Self` type against the corresponding field
+in the `Target` type. Assuming `Fs` is the type of a field in `Self` and `Ft` is
+the type of the corresponding field in `Target`, then either `Ft <: Fs` or
+`Fs: CoerceUnsized<Ft>` (note that this includes built-in coercions).
```

Comment:
How does this include builtin coercions? There are some coercions, at least, that are not represented by the `CoerceUnsized` trait, right? I guess that's primarily the coercions around fn pointers and in particular `unsafe` fn items.